python/ray/util/annotations.py

```diff
@@ -107,22 +113,52 @@ def Deprecated(*args, **kwargs):
     if len(args) == 1 and len(kwargs) == 0 and callable(args[0]):
         return Deprecated()(args[0])
 
-    message = (
-        "**DEPRECATED:** This API is deprecated and may be removed in a future "
-        "Ray release."
+    doc_message = (
+        "\n    DEPRECATED: This API is deprecated and may be removed "
+        "in future Ray releases."
+    )
+    warning_message = (
+        "This API is deprecated and may be removed in future Ray releases. "
+        "You could suppress this warning by setting env variable "
+        "RAY_DISABLE_DEPRECATION_WARNING=0."
     )
 
+    warning = kwargs.pop("warning", False)
+
     if "message" in kwargs:
-        message += " " + kwargs["message"]
+        doc_message = doc_message + " " + kwargs["message"]
+        warning_message = warning_message + " " + kwargs["message"]
         del kwargs["message"]
 
     if kwargs:
         raise ValueError("Unknown kwargs: {}".format(kwargs.keys()))
 
     def inner(obj):
-        _append_doc(obj, message=message, directive="warning")
+        _append_doc(obj, message=doc_message, directive="warning")
         _mark_annotated(obj)
-        return obj
+
+        if not warning:
+            return obj
+
+        if inspect.isclass(obj):
+            obj_init = obj.__init__
+
+            def patched_init(*args, **kwargs):
+                if os.environ.get(RAY_DISABLE_DEPRECATION_WARNING, "0") != "1":
```

Comment:
this prints warning every time the API is called. Should we do it only once instead?

Reply:
hmm i thought the default behavior prints the first occurrence of matching warnings for each location (module + line number) where the warning is issued

https://docs.python.org/3/library/warnings.html#the-warnings-filter